aleph/logic/extractors/extract.py

```diff
@@ -1,71 +1,51 @@
-import os
-import spacy
 import logging
-from polyglot.text import Text
+from alephclient.services.common_pb2 import Text
+from alephclient.services.entityextract_pb2 import ExtractedEntity
+from alephclient.services.entityextract_pb2_grpc import EntityExtractStub
 
 from aleph import settings
+from aleph.services import ServiceClientMixin
 from aleph.logic.extractors.result import PersonResult, LocationResult
 from aleph.logic.extractors.result import OrganizationResult, LanguageResult
+from aleph.util import backoff
 
 log = logging.getLogger(__name__)
 
-MIN_LENGTH = 60
-MAX_LENGTH = 100000
 
-POLYGLOT_PATH = os.environ.get('POLYGLOT_DATA_PATH')
-POLYGLOT_NER_PATH = os.path.join(POLYGLOT_PATH, 'polyglot_data/ner2')
-POLYGLOT_LANGUAGES = os.listdir(POLYGLOT_NER_PATH)
-POLYGLOT_TYPES = {
-    'I-PER': PersonResult,
-    'I-ORG': OrganizationResult,
-    'I-LOC': LocationResult
-}
+class NERService(ServiceClientMixin):
+    SERVICE = settings.NER_SERVICE
+    MIN_LENGTH = 60
+    MAX_LENGTH = 100000
+    TYPES = {
+        ExtractedEntity.ORGANIZATION: OrganizationResult,
+        ExtractedEntity.PERSON: PersonResult,
+        ExtractedEntity.LOCATION: LocationResult,
+        ExtractedEntity.LANGUAGE: LanguageResult
+    }
 
-# https://spacy.io/api/annotation#named-entities
-SPACY_TYPES = {
-    'PER': PersonResult,
-    'PERSON': PersonResult,
-    'ORG': OrganizationResult,
-    'LOC': LocationResult,
-    'GPE': LocationResult
-}
-
-
-def extract_polyglot(agg, text, languages):
-    if not settings.ENABLE_POLYGLOT:
-        return
-    if len(text) < MIN_LENGTH:
-        return
-    try:
-        parsed = Text(text)
-        lang = parsed.language
-        if lang.confidence > 90:
-            yield LanguageResult.create(agg, lang.code, None, None)
-        if lang.code not in POLYGLOT_LANGUAGES:
+    def extract(self, text, languages):
+        if text is None or len(text) < self.MIN_LENGTH:
             return
-        for entity in parsed.entities:
-            label = ' '.join(entity)
-            clazz = POLYGLOT_TYPES.get(entity.tag)
-            if clazz is not None:
-                yield clazz.create(agg, label, entity.start, entity.end)
-    except Exception as ex:
-        log.warning("Polyglot failed: %s" % ex)
-
-
-def extract_spacy(agg, text, languages):
-    if not settings.ENABLE_SPACY:
-        return
-    if len(text) < MIN_LENGTH:
-        return
-    try:
-        text = text[:MAX_LENGTH]
-        if not hasattr(settings, '_spacy_model'):
-            log.debug("Loading spaCy NER model...")
-            settings._spacy_model = spacy.load('xx')
-        doc = settings._spacy_model(text)
-        for ent in doc.ents:
-            clazz = SPACY_TYPES.get(ent.label_)
-            if clazz is not None:
-                yield clazz.create(agg, ent.text, ent.start, ent.end)
-    except Exception:
-        log.exception("spaCy failed")
+        text = text[:self.MAX_LENGTH]
```

Comment:
Why do we need to trim the text? Just to keep the memory footprint smaller or is there any other reason?

Reply:
spaCy has a limit, Polyglot doesn't. gRPC also limits out around about 4MiB. Do you think this should be done in the server?

Reply:
If grpc has a limit, then we have to enforce that limit before shoving it down the wire though. How about sending it in chunks instead of throwing the rest of the text away though?